### ISSUE TITLE
Session how big should be the scope

### ISSUE THREAD

**Issue by BoasE** [neo4j/neo4j-dotnet-driver#70]
When looking at the conecpt of a session. how long is it intended to exists?

for example when I import a lot of data, would I create one session covering the whole process ? Or should I create a session for each import step (e.g. each kind of node).

Furthermore is a session meant to be threadsafe?

As comparison I currently use the dotnet client nuget package and there I create one transaction for the whole impor process to have better performance, there for I would assume to get one session for the import process. but as this works currently in a async way the session should be threadsafe or I need to create some kind of producer / consumer scenario

**Comment by zhenlineo** [comment]
Session is intended to be a long running connection with the server. The session is also designed to be re-useable. When we do `session.Dispose` is will not close the connection with the server but instead it would try to buffer the connection with the server by returning it to the `idle session pool`. So that is to say, it is really cheap to dispose the session. The max number of the sessions that will be buffered by this irks session pool could be configured via `Config`.

The session/the whole driver currently is implemented with blocking operations, which means if the `session.Run(...).Consume()` returned then we ensure that the cypher has already executed and all results has returned.

Let's see an example. Say thread1 and thread2 are sharing a session. Thread1 wants to run cypher statement A, B, C. Similarly thread2 wants to run F, G, H. 

Thread1: A, B, C.
Thread2: F, G, H.

Then we just run them concurrently. 
The session would guarantee the order ABC and FGH. But it will not guarantee any order between them. So it could be ABCFGH, it could also be AFGBHC. 

However as I have said, we have a idle session pool and we will reuse idle sessions. It is super cheap to dispose/close the session when you do not need it to return it to idle session pool. 

So I would suggest you just share a `driver` among threads and get a `new` session each time your thread needs it.

**Comment by zhenlineo** [comment]
And of course, dispose the session after your thread done with it to return it to the pool so that other threads could reclaim it.

**Comment by zhenlineo** [comment]
Reading your question again. The session is not designed to be thread-safe. If you want to run concurrent operations against the database, simply create multiple sessions objects. Sharing a session in two threads is a bad idea. 

Even if right now the session will run the cyphers sequentially, but your thread has to make sure no nested tx in this session. 

So it is strongly recommended to use multiple sessions for running concurrent operations. As long as you close the session (one and only once) in each thread, then the driver will manage the sessions for you.
The code will also be clean for you as it just looks like the following in each of your thread working with db:

```
using(var session = driver.Session())
{
    var result = session.Run("CREATE (n) RETURN n");
    then do something with the result
}
```

**Comment by zhenlineo** [comment]
I am closing this issue as I am expecting your question is already answered. Pls be free to re-open this issue or raise another issue if you still get questions.
